File: wrappers/java/src/test/java/org/hyperledger/indy/sdk/signus/DecryptTest.java
package org.hyperledger.indy.sdk.signus;

import org.hyperledger.indy.sdk.IndyIntegrationTestWithPoolAndSingleWallet;
import org.hyperledger.indy.sdk.InvalidStructureException;
import org.hyperledger.indy.sdk.signus.SignusResults.CreateAndStoreMyDidResult;
import org.hyperledger.indy.sdk.wallet.WalletValueNotFoundException;
import org.junit.Before;
import org.junit.Test;

import java.util.Arrays;
import java.util.concurrent.ExecutionException;

import static org.hamcrest.CoreMatchers.isA;
import static org.junit.Assert.assertTrue;

public class DecryptTest extends IndyIntegrationTestWithPoolAndSingleWallet {

	private String trusteeDid;
	private String myDid;
	private String myVerkey;

	@Before
	public void before() throws Exception {
		CreateAndStoreMyDidResult trusteeNym = Signus.createAndStoreMyDid(wallet, TRUSTEE_IDENTITY_JSON).get();
		trusteeDid = trusteeNym.getDid();
		String trusteeVerkey = trusteeNym.getVerkey();

		CreateAndStoreMyDidResult myNym = Signus.createAndStoreMyDid(wallet, MY1_IDENTITY_JSON).get();
		myDid = myNym.getDid();
		myVerkey = myNym.getVerkey();

		String identityJson = String.format(IDENTITY_JSON_TEMPLATE, trusteeDid, trusteeVerkey);
		Signus.storeTheirDid(wallet, identityJson).get();
	}

	@Test
	public void testDecryptWorks() throws Exception {
		byte[] decryptedMessage = Signus.decrypt(wallet, pool, myDid, trusteeDid, ENCRYPTED_MESSAGE, NONCE).get();
		assertTrue(Arrays.equals(MESSAGE, decryptedMessage));
	}

	@Test
	public void testDecryptWorksForOtherCoder() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		String identityJson = String.format(IDENTITY_JSON_TEMPLATE, myDid, myVerkey);
		Signus.storeTheirDid(wallet, identityJson).get();

		SignusResults.EncryptResult encryptResult = Signus.encrypt(wallet, pool, myDid, myDid, MESSAGE).get();

		Signus.decrypt(wallet, pool, myDid, trusteeDid, encryptResult.getEncryptedMessage(), encryptResult.getNonce()).get();
	}

	@Test
	public void testDecryptWorksForNonceNotCorrespondMessage() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(InvalidStructureException.class));

		byte[] nonce = {46, 33, - 4, 67, 1, 44, 57, - 46, - 91, 87, 14, 41, - 39, 48, 42, - 126, - 121, 84, - 58, 59, - 27, 51, - 32, - 23};

		Signus.decrypt(wallet, pool, myDid, trusteeDid, ENCRYPTED_MESSAGE, nonce).get();
	}

	@Test
	public void testDecryptWorksForUnknownMyDid() throws Exception {
		thrown.expect(ExecutionException.class);
		thrown.expectCause(isA(WalletValueNotFoundException.class));

		Signus.decrypt(wallet, pool, DID, trusteeDid, ENCRYPTED_MESSAGE, NONCE).get();
	}
}

File: wrappers/dotnet/indy-sdk-dotnet-test/SignusTests/DecryptTest.cs
using Hyperledger.Indy.SignusApi;
using Hyperledger.Indy.WalletApi;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using System;
using System.Linq;
using System.Threading.Tasks;

namespace Hyperledger.Indy.Test.SignusTests
{
    [TestClass]
    public class DecryptTest : IndyIntegrationTestWithPoolAndSingleWallet
    {
        private string _trusteeDid;
        private string _trusteeVerkey;
        private string _myDid;
        private string _myVerkey;
        private byte[] _encryptedMessage = (byte[]) (Array) new sbyte[] { -105, 30, 89, 75, 76, 28, -59, -45, 105, -46, 20, 124, -85, -13, 109, 29, -88, -82, -8, -6, -50, -84, -53, -48, -49, 56, 124, 114, 82, 126, 74, 99, -72, -78, -117, 96, 60, 119, 50, -40, 121, 21, 57, -68, 89 };
        private byte[] _nonce = (byte[])(Array) new sbyte[] { -14, 102, -41, -57, 1, 4, 75, -46, -91, 87, 14, 41, -39, 48, 42, -126, -121, 84, -58, 59, -27, 51, -32, -23 };

        [TestInitialize]
        public async Task CreateWalletWithDid()
        {           
            var trusteeNym = await Signus.CreateAndStoreMyDidAsync(wallet, TRUSTEE_IDENTITY_JSON);
            _trusteeDid = trusteeNym.Did;
            _trusteeVerkey = trusteeNym.VerKey;

            var myNym = await Signus.CreateAndStoreMyDidAsync(wallet, MY1_IDENTITY_JSON);
            _myDid = myNym.Did;
            _myVerkey = myNym.VerKey;

            var identityJson = string.Format(IDENTITY_JSON_TEMPLATE, _trusteeDid, _trusteeVerkey);
            await Signus.StoreTheirDidAsync(wallet, identityJson);
        }
               
        [TestMethod]
        public async Task TestDecryptWorks()
        {
            var decryptedMessage = await Signus.DecryptAsync(wallet, _myDid, _trusteeDid, _encryptedMessage, _nonce);
            Assert.IsTrue(MESSAGE.SequenceEqual(decryptedMessage));
        }

        [TestMethod]
        public async Task TestDecryptWorksForOtherCoder()
        {
            var identityJson = string.Format(IDENTITY_JSON_TEMPLATE, _myDid, _myVerkey);
            await Signus.StoreTheirDidAsync(wallet, identityJson);

            var encryptResult = await Signus.EncryptAsync(wallet, pool, _myDid, _myDid, MESSAGE);            

            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
                Signus.DecryptAsync(wallet, _myDid, _trusteeDid, encryptResult.EncryptedMsg, encryptResult.Nonce)
            );;
        }

        [TestMethod]
        public async Task TestDecryptWorksForNonceNotCorrespondMessage()
        {
            var nonce = (byte[])(Array)new sbyte[] { 46, 33, -4, 67, 1, 44, 57, -46, -91, 87, 14, 41, -39, 48, 42, -126, -121, 84, -58, 59, -27, 51, -32, -23 };
                        
            var ex = await Assert.ThrowsExceptionAsync<InvalidStructureException>(() =>
               Signus.DecryptAsync(wallet, _myDid, _trusteeDid, _encryptedMessage, nonce)
            );
        }

        [TestMethod]
        public async Task TestDecryptWorksForUnknownMyDid()
        {
            var ex = await Assert.ThrowsExceptionAsync<WalletValueNotFoundException>(() =>
               Signus.DecryptAsync(wallet, "unknowDid", _trusteeDid, _encryptedMessage, _nonce)
           );
        }
    }
}
